ecosystem/sep-0010.md

```diff
@@ -199,6 +204,8 @@ Upon successful verification, **Server** responds with a session JWT, containing
 * `iat` (the time at which the JWT was issued [RFC7519, Section 4.1.6](https://tools.ietf.org/html/rfc7519#section-4.1.6)) — current timestamp (`1530644093`)
 * `exp` (the expiration time on or after which the JWT must not be accepted for processing, [RFC7519, Section 4.1.4](https://tools.ietf.org/html/rfc7519#section-4.1.4)) — a server can pick its own expiration period for the token (`1530730493`)
 * `client_domain` - (optional) a nonstandard JWT claim containing the client home domain, included if the challenge transaction contained a `client_domain` (see [Verifying Client Application Identity](#verifying-client-application-identity))
+* `memo_type` — (optional) the memo type provided in the challenge, one of `text`, `id` or `hash`. See the [Memos](#memos) section for more information.
+* `memo` — (optional) the memo provided in the challenge, always of type `string` (even when `memo_type` is equal to `id`) to support parsing the value in languages that don't support big numbers.
```

Comment:
❗ One issue we have with making the JWT's identity a sum of `sub`, `memo_type`, and `memo`, is that it is not consistent with [RFC7519 Section 4.1.2] which states that that the `sub` **MUST** uniquely identify the principal being identified.

>The "sub" (subject) claim identifies the principal that is the
>subject of the JWT.  The claims in a JWT are normally statements
>about the subject.  The subject value MUST either be scoped to be
>locally unique in the context of the issuer or be globally unique.
>The processing of this claim is generally application specific.

-- [RFC7519 Section 4.1.2]

[RFC7519 Section 4.1.2]: https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.2

In the past I think we discussed other designs with we try to encode the memo and memo type into the `sub`, but I don't remember where we landed on that. The `sub` claim must be a string or a URI so unfortunately we can't use a JSON object. We could reconsider a `sub` value in the form `<account>,<memotype>,<memo>`, or similar.

Reply:
Changing the format of the `sub` string means that users of the `sub` field need to update their parsing logic. In most cases, the only service using the decoded contents of the JWT is the same service that issued the JWT, but I can imagine cases where a service or client has not updated to parse the new format.

Personally I would rather deviate from the RFC spec and the `sub` format unchanged but I don't feel strongly about it.

One question I have about this change is, how would non-shared accounts be represented? Would there be separators in the `sub` string still? I would vote that if we make this change, the separators should always be present:
```
# with "sub" --> "G...,," or "G...,123,id"
>>> account, memo, memo_type = jwt.get("sub").split(",")
```

Otherwise, we'll have to implement logic like so:
```
# with "sub" --> "G..." or "G...,123,id"
>>> account, memo, memo_type = None, None, None
>>> if "," in jwt.get("sub"):
>>>     account, memo, memo_type = jwt.get("sub")
>>> else:
>>>     account = jwt.get("sub")
```

Reply:
> I can imagine cases where a service or client has not updated to parse the new format

Even with the existing proposal services that aren't updated are likely to contain bugs, since the only way to accurately identify the user is to consume the `sub`, `memo_type`, and `memo` claims. In some ways changing the subject to ensure that the full identity of the user is captured within it may be more resilient to a more catastrophic bug.

---

To answer the question about how this would be implemented, I would take the approach of assuming they could always be present, but ignoring them if they are not:

```go
var account, memo_type, memo string
subParts := strings.SplitN(sub, ",", 3)
account = subParts[0]
if len(subParts) > 1 {
    if len(subParts) < 3 {
        return fmt.Errorf("invalid number of components to jwt subject: %q", sub)
    }
    memo_type = subParts[1]
    memo = subParts[2]
}
```

Reply:
Thats a good point about un-updated services, and the code above is certainly more resilient. 👍 

I'll update the PR accordingly. Just to make sure: this format should only be used in the `sub` field of the JWT right? Request parameters remain unchanged?

Reply:
Another way to interpret the RFC is to say that the principal is the shared account holder, who is requesting authentication for one of their users. It's odd because from an applications point-of-view the subject is usually the user, although from a key management perspective it is somewhat accurate because the key holder is the account holder.

Reply:
> I'll update the PR accordingly. Just to make sure: this format should only be used in the `sub` field of the JWT right? Request parameters remain unchanged?

Yes request parameters can remain unchanged.

I wasn't really meaning to propose this format, but rather highlight maybe we could do something like this. We should maybe consider if there are better formats, although comma-separated is simple to implement.

Another option is to encode the memo into the muxed account strkey format, but that will make it not possible to distinguish muxed accounts from memo accounts in these setups which might not be great.

Reply:
Thats also true, although I think `sub` needs fully represent the scope of the session based on the specification's definition, and the scope would not be the same for two JWT's that use the same `sub` value but have different `memo` and `memo_type` parameters. 

> The claims in a JWT are normally statements about the subject.

You also have to squint hard to think that the `memo` and `memo_type` claims are _about_ the shared account holder, since they're really changing the identity of the subject, since (account, null, null) and (account, 123, text) describe two distinct entities from the anchor's perspective.

Reply:
I've updated the `sub` format to be `G...:1234567` in the case of a shared account.